### PR TITLE
Integrate libsodium v1.0.18 as a component.

### PR DESCRIPTION
**Pull request by aemseemann**
This PR adds the [sodium](https://github.com/jedisct1/libsodium) library of cryptographic functions to Sming as discussed in https://github.com/SmingHub/Sming/issues/1781

See Sming/Components/libsodium/Readme.md for further information.

Aside from the library, I fixed the partition table generation in startup.cpp for devices with flash sizes <= 1MB (should fix the issue reported in https://github.com/SmingHub/Sming/issues/1570#issuecomment-466719355) and added a definition of `__errno`, which does not exist in libc for some reasons, but is referenced by some functions in libsodium (via `errno`).

The

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -118,3 +118,8 @@
 	path = Sming/Libraries/MCP_CAN_lib
 	url = https://github.com/coryjfowler/MCP_CAN_lib.git
 	ignore = dirty
+[submodule "Sming/Components/libsodium/libsodium"]
```
Comment (slaff) [inline .gitmodules]: Please, place this submodule definition block after esptool2. Also change the fist line to:
`[submodule "libsodium"]`. This way looking at the gitmodules file the developers can easier figure out that this is a submodule that can be used in all architectures and is not an Arduino-style library.

File: Sming/Components/libsodium/component.mk [hunk 2]
```diff
@@ -0,0 +1,103 @@
+# umm_malloc (custom heap allocation)
```
Comment (mikee47) [inline Sming/Components/libsodium/component.mk]: comment

File: .gitmodules (hunk 1 above)
Comment (mikee47) [inline .gitmodules]: Should go at top (e.g. line 24) and rename to "libsodium"

File: Sming/Components/libsodium/Readme.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# libsodium
```
Comment (mikee47) [inline Sming/Components/libsodium/Readme.md]: Please rename file to `README.md` so doc. system picks it up.

File: .gitmodules (hunk 1 above)
Comment (slaff) [reply .gitmodules]: Another point. The code at the moment does not compile for the host architecture. You can repeat the following steps to reproduce the issue:

```
cd $SMING_HOME/../samples/HttpServer_FirmwareUpload
echo "COMPONENT_DEPENDS := libsodium" >> component.mk
make SMING_ARCH=Host ENABLE_GDB=1 ENABLE_LWIPDEBUG=1 ENABLE_MALLOC_COUNT=0
```

The compilation errors with:
```
Sming/Components/libsodium/libsodium/src/libsodium/randombytes/sysrandom/randombytes_sysrandom.c:30:31: error: implicit declaration of function ‘syscall’ [-Werror=implicit-function-declaration]
 #   define getrandom(B, S, F) syscall(SYS_getrandom, (B), (int) (S), (F))
                               ^
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/libsodium/libsodium/src/libsodium/randombytes/sysrandom/randombytes_sysrandom.c:265:18: note: in expansion of macro ‘getrandom’
         readnb = getrandom(buf, size, 0);
                  ^
cc1: all warnings being treated as errors
```

It would be great if `libsodium` works also on the Host architecture.

File: Sming/Components/libsodium/component.mk [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	$(LIBSODIUM_SRC)/crypto_stream/salsa2012/stream_salsa2012.c \
+	$(LIBSODIUM_SRC)/crypto_stream/salsa208/ref/stream_salsa208_ref.c \
+	$(LIBSODIUM_SRC)/crypto_stream/salsa208/stream_salsa208.c \
+	$(LIBSODIUM_SRC)/crypto_stream/xchacha20/stream_xchacha20.c \
+	$(LIBSODIUM_SRC)/randombytes/sysrandom/randombytes_sysrandom.c
+
+COMPONENT_INCDIRS := $(LIBSODIUM_ROOT)/src/libsodium/include
+
+# results from configure script run for --host=xtensa-lx106-elf (unused defines removed)
+COMPONENT_CFLAGS := -DSODIUM_STATIC -DHAVE_C_VARARRAYS=1 -DNATIVE_LITTLE_ENDIAN=1 -DHAVE_INLINE_ASM=1 -DHAVE_WEAK_SYMBOLS=1 -DHAVE_ALLOCA_H=1 -DHAVE_ALLOCA=1 -DCONFIGURED=1
```
Comment (slaff) [inline Sming/Components/libsodium/component.mk]: Adding  -D_GNU_SOURCE=1 to the list of COMPONENT_CFLAGS makes the compilation work also under Host (Linux in my case).

File: Sming/Arch/Esp8266/Components/esp8266/startup.cpp [hunk 5]
```diff
@@ -106,7 +106,7 @@ extern "C" void ICACHE_FLASH_ATTR WEAK_ATTR user_pre_init(void)
 	static const partition_item_t partitions[] = {
 			PARTITION_ITEM(SYSTEM_PARTITION_BOOTLOADER,			0,					1),
 			PARTITION_ITEM(SYSTEM_PARTITION_RBOOT_CONFIG,		1,					1),
-			PARTITION_ITEM(SYSTEM_PARTITION_PROGRAM,			2,					MAX_PROGRAM_SECTORS - 2),
+			PARTITION_ITEM(SYSTEM_PARTITION_PROGRAM,			2,					(rfCalSector < MAX_PROGRAM_SECTORS ? rfCalSector : MAX_PROGRAM_SECTORS) - 2),
```
Comment (slaff) [inline Sming/Arch/Esp8266/Components/esp8266/startup.cpp]: Revert this change. PR #1800 will take care of it.

**Comment by mikee47** [comment]
Can we have a small, simple sample to check this build?

**Comment by slaff** [comment]
> Can we have a small, simple sample to check this build?

The HttpServer_FirmwareUpload sample can be extended to support signed ROMS.

File: Sming/Arch/Esp8266/Components/esp8266/startup.cpp [hunk 6]
```diff
@@ -134,3 +134,11 @@ namespace std {
         while(1);
     };
 }
+
+extern "C" {
+    __attribute__((weak)) int *__errno(void)
```
Comment (mikee47) [inline Sming/Arch/Esp8266/Components/esp8266/startup.cpp]: Don't need this, add `#include <sys/errno.h>` somewhere.
(See `Arch/Esp8266/Components/gdbstub/gdbstub.cpp`.)

File: Sming/Components/libsodium/component.mk [hunk 7]
```diff
@@ -0,0 +1,103 @@
+# umm_malloc (custom heap allocation)
+
+COMPONENT_SUBMODULES	:= libsodium
+
+LIBSODIUM_ROOT := $(COMPONENT_PATH)/libsodium
+LIBSODIUM_SRC := $(LIBSODIUM_ROOT)/src/libsodium
+
+COMPONENT_SRCFILES := \
```
Comment (mikee47) [inline Sming/Components/libsodium/component.mk]: If you wanted to avoid listing all files individually, see `Sming/component.mk` for use of ListAllSubDirs.

Comment (aemseemann) [reply Sming/Components/libsodium/component.mk]: In this case the file listing is intentional, because I excluded all files with architecture-specific assembly-optimized implementations.

File: .gitmodules (hunk 1 above)
Comment (aemseemann) [reply .gitmodules]: I put it under esptool2, as slaff proposed.

File: Sming/Components/libsodium/component.mk (hunk 2 above)
Comment (aemseemann) [reply Sming/Components/libsodium/component.mk]: removed leftover comment

File: Sming/Components/libsodium/Readme.md (hunk 3 above)
Comment (aemseemann) [reply Sming/Components/libsodium/Readme.md]: done

File: .gitmodules (hunk 1 above)
Comment (aemseemann) [reply .gitmodules]: It was also my intention to make it available to all supported architectures. 
Adding _GNU_SOURCE fix the issue for me, too. However, It would be great if someone can check this out on a Windows host.
Offering a working solution for all possible realizations of the Host "architecture" can get quite involved, as it is not even restricted to x86-based Hosts.

File: Sming/Arch/Esp8266/Components/esp8266/startup.cpp (hunk 5 above)
Comment (aemseemann) [reply Sming/Arch/Esp8266/Components/esp8266/startup.cpp]: Done.

File: Sming/Arch/Esp8266/Components/esp8266/startup.cpp (hunk 6 above)
Comment (aemseemann) [reply Sming/Arch/Esp8266/Components/esp8266/startup.cpp]: This header is already included by <errno.h>. I don't see how including it explicitly should fix the linker error. At least in my toolchain built, sys/errno.h contains the following:

```
#define errno (*__errno())
extern int *__errno _PARAMS ((void));
```

I.e. to use errno, a definition of the `__errno` function has to be provided somewhere. Naturally, it should be in libc, but since Sming uses a stripped-down version of libc, it is not there, which is why I added it to startup.cpp (seemed the logical place, since it is a similar situation with the missing `__throw_bad_function_call`).

Please advise.

**Comment by aemseemann** [comment]
I'll rebase on current develop and work on the example now.

File: Sming/Arch/Esp8266/Components/esp8266/startup.cpp (hunk 6 above)
Comment (mikee47) [reply Sming/Arch/Esp8266/Components/esp8266/startup.cpp]: Strange, it seems `errno` isn't actually used anywhere else (just the codes).
`libmicroc` has a definition for `__errno_location` which should probably be referenced in a header somewhere, but it isn't.

https://github.com/anakod/esp_microc/blob/master/errno/__errno_location.c

I'll raise another PR for this.

File: Sming/Components/libsodium/.patches/libsodium.patch [hunk 8]
```diff
@@ -0,0 +1,1232 @@
...
++randombytes_sysrandom_stir(void)
++{
++}
++
++static void
++randombytes_sysrandom_buf(void * const buf, const size_t size)
++{
++    uint8_t *dst = buf;
++    for (size_t remaining = size; remaining > 0; --remaining) {
++        *dst++ = (uint8_t)(RANDOM_REG32 & 0xFF);
```
Comment (mikee47) [inline Sming/Components/libsodium/.patches/libsodium.patch]: I think `os_get_random()` would be a better choice.

File: Sming/Components/libsodium/.patches/libsodium.patch [hunk 9]
```diff
@@ -0,0 +1,1232 @@
...
++    return RANDOM_REG32;
++}
++
++static void
++randombytes_sysrandom_stir(void)
++{
++}
++
++static void
++randombytes_sysrandom_buf(void * const buf, const size_t size)
```
Comment (mikee47) [inline Sming/Components/libsodium/.patches/libsodium.patch]: I think os_get_random() would be a better choice. It's also available in the emulator (and ESP32).

File: Sming/Components/libsodium/.patches/libsodium.patch [hunk 10]
```diff
@@ -0,0 +1,1232 @@
...
+ #include "randombytes.h"
+@@ -108,7 +112,35 @@ randombytes_sysrandom_close(void)
+     return 0;
+ }
+ 
+-#else /* HAVE_SAFE_ARC4RANDOM */
++#elif defined(ARCH_ESP8266)
++
++static uint32_t
++randombytes_sysrandom(void)
```
Comment (mikee47) [inline Sming/Components/libsodium/.patches/libsodium.patch]: I think `os_random()` would be better. 

`os_random()` calls `r_rand()`, which calls `phy_get_rand()`. This reads `RANDOM_REG32` but also ensures the clock is enabled.

os_random() is provided in the emulator and ESP32. More background:

https://raw.githubusercontent.com/pfalcon/esp8266-re-wiki-mirror/master/Random_Number_Generator.mw

Note that lwip uses `r_rand()` directly to save an additional call. I wouldn't bother as we shouldn't be calling it in a loop anyway.

File: Sming/Arch/Esp8266/Components/esp8266/startup.cpp (hunk 6 above)
Comment (aemseemann) [reply Sming/Arch/Esp8266/Components/esp8266/startup.cpp]: Ok, I'll rebase again when https://github.com/SmingHub/Sming/pull/1801 is merged and then remove the definition in startup.cpp

File: Sming/Components/libsodium/.patches/libsodium.patch (hunk 10 above)
Comment (mikee47) [reply Sming/Components/libsodium/.patches/libsodium.patch]: I'll do a PR for these.

Comment (aemseemann) [reply Sming/Components/libsodium/.patches/libsodium.patch]: Thanks for the information. I wasn't aware of the details and that the HWRNG is already used by the SDK. I replaced the calls as you suggested and also added your link to the documentation.

File: Sming/Arch/Esp8266/Components/esp8266/startup.cpp (hunk 6 above)
Comment (slaff) [reply Sming/Arch/Esp8266/Components/esp8266/startup.cpp]: > Ok, I'll rebase again when #1801 is merged

PR #1801 is merged in develop.

**Comment by aemseemann** [comment]
I integrated signature verification into the HttpServer_FirmwareUpload example. 
It works on my ESP01 board, but I would still consider it work in progress.
The life cycle management for streams created in `fileUploadMapper` is kind of a mess. My debug logs suggest that streams are never destroyed, causing a potentially massive memory leak. Also, stream methods `init()` and `close()` are never invoked. The latter would be invoked from the destructor, but this probably is *after* the request completion callback (i.e. the `onUpload` method in this example) has returned. 

Second issue: Travis CI does not have the necessary python package to create the signing and verification keys. Can this package be installed or do you have any suggestion for a workaround?

**Comment by mikee47** [comment]
Re. travis If you look in `.travis/install.sh` you can add required python packages in there.

**Comment by aemseemann** [comment]
Travis issue fixed, thanks for the hint. Learned a lot today!

**Comment by mikee47** [comment]
I'm having issues using the signtool with python 3, is there an updated version? Also, you could put the tool plus all of the make logic into the libsodium Component. Take a look at `Arch/Esp8266/Components/axtls-8266` which does something similar.

**Comment by mikee47** [comment]
Patch for python 3. Breaks python 2 though :-(

```
diff --git a/samples/HttpServer_FirmwareUpload/signtool.py b/samples/HttpServer_FirmwareUpload/signtool.py
index 0a0562ff8..ff43b3331 100755
--- a/samples/HttpServer_FirmwareUpload/signtool.py
+++ b/samples/HttpServer_FirmwareUpload/signtool.py
@@ -39,5 +39,5 @@ extern "C" {
 
 static const uint8_t firmwareVerificationKey[''' + str(len(pk)) + '''] = {
-    ''' + ', '.join('0x%02X' % ord(b) for b in pk) + '''
+    ''' + ', '.join('0x%02X' % b for b in pk) + '''
 };
 #ifdef __cplusplus
@@ -109,5 +109,5 @@ def main():
         if args.keyfile:                
             with open(args.keyfile, 'wb') as privkeyfile:
-                privkeyfile.write(sk_seed.encode('Hex'))
+                privkeyfile.write(sk_seed)
             print("Private signing key written to '%s'." % args.keyfile)
         else:
@@ -115,6 +115,6 @@ def main():
     elif args.keyfile:
         print("Read signing key from '%s'" % args.keyfile)
-        with open(args.keyfile, 'r') as keyfile:
-            sk_seed = keyfile.read().decode('Hex')
+        with open(args.keyfile, 'rb') as keyfile:
+            sk_seed = keyfile.read()
     
     else:
```

**Comment by mikee47** [comment]
We should be focused on Python 3 anyway - https://python3statement.org/

**Comment by mikee47** [comment]
@aemseemann I've pushed some fixes to your repo https://github.com/aemseemann/Sming/pull/2
Builds for all arches, and runs in host emulator on both Linux and Windows.

**Comment by aemseemann** [comment]
@mikee47: I specifically wrote signtool.py for python2, because the other tools in the build process (esptool.py, make terminal) also require python2 (which, btw, is a PITA for users of distros where "python" defaults to python3). I am very aware of the python2 deprecation notice and agree with you that Sming should move to python3 sooner than later. However, I would avoid a situation where both python2 and python3 is needed in the same build process. Please correct me if I misunderstood something.

> Also, you could put the tool plus all of the make logic into the libsodium Component.
This seems odd to me, because firmware signing is no feature of the libsodium component. I would rather create a new component that contains all the OTA upgrade stuff (classes like `SignedRbootOutputStream`, the signtool, make targets, and so on) and depends on libsodium. But IMO this is also content for a separate PR.

**Comment by mikee47** [comment]
I'll raise an issue concerning Python as we need to agree some guidelines. For info, I'm running with python3 as default (linux 3.6.8, Windows 3.7.3) and both esptool and pyterm work fine.

> This seems odd to me, because firmware signing is no feature of the libsodium component. I would rather create a new component that contains all the OTA upgrade stuff (classes like `SignedRbootOutputStream`, the signtool, make targets, and so on) and depends on libsodium. But IMO this is also content for a separate PR.

Agreed.

**Comment by aemseemann** [comment]
> For info, I'm running with python3 as default

My bad, I simply forgot to install PySerial and did not read the error messages carefully enough. Runs fine for me now, too.

File: samples/HttpServer_FirmwareUpload/app/application.cpp [hunk 11]
```diff
@@ -22,7 +132,7 @@ void onFile(HttpRequest& request, HttpResponse& response)
 	}
 
 	String file = request.uri.getRelativePath();
-
+	Serial.printf("Requested file: %s\n", file.c_str());
```
Comment (slaff) [inline samples/HttpServer_FirmwareUpload/app/application.cpp]: Please, remove this line.

File: samples/HttpServer_FirmwareUpload/app/application.cpp [hunk 12]
```diff
@@ -37,44 +147,35 @@ void onFile(HttpRequest& request, HttpResponse& response)
 
 int onUpload(HttpServerConnection& connection, HttpRequest& request, HttpResponse& response)
 {
+	Serial.printf("onUpload\n");
```
Comment (slaff) [inline samples/HttpServer_FirmwareUpload/app/application.cpp]: If you want to use debug code don't use `Serial.printf` but `debug_d` instead.

File: samples/HttpServer_FirmwareUpload/app/application.cpp [hunk 13]
```diff
@@ -37,44 +147,35 @@ void onFile(HttpRequest& request, HttpResponse& response)
...
-	uint8_t slot = bootConfig.current_rom;
-	slot = (slot == 0 ? 1 : 0);
-	Serial.printf("Firmware updated, rebooting to rom %d...\r\n", slot);
-	rboot_set_current_rom(slot);
-	System.restart(5); // defer the restart with 5 seconds to give time to the web server to return the response
-
-	response.sendFile("restart.html");
-
+	SignedRbootOutputStream* uploadStream = static_cast<SignedRbootOutputStream*>(file);
+	uploadStream->close();
```
Comment (slaff) [inline samples/HttpServer_FirmwareUpload/app/application.cpp]: Don't close the stream here. It will be closed later at the end of the request.

File: samples/HttpServer_FirmwareUpload/app/application.cpp [hunk 14]
```diff
@@ -3,10 +3,120 @@
 #include <Data/Stream/LimitedMemoryStream.h>
 #include <MultipartParser/HttpMultipartResource.h>
 #include <Data/Stream/RbootOutputStream.h>
+#include <sodium/crypto_sign.h>
+#include "FirmwareVerificationKey.h"
 
 HttpServer server;
 String lastModified;
 
+class SignedRbootOutputStream : public RbootOutputStream
```
Comment (slaff) [inline samples/HttpServer_FirmwareUpload/app/application.cpp]: Take that class out of  application.cpp. You can move it to a separate file in this sample OR if you think it is good enough to be reused move it to `Components/libsodium` . Similar to [MultipartParser](https://github.com/SmingHub/Sming/tree/develop/Sming/Components/MultipartParser). I would recommend you to do the latter.

File: samples/HttpServer_FirmwareUpload/app/application.cpp [hunk 15]
```diff
@@ -3,10 +3,120 @@
...
+{
+	const uint8_t* const verificationKey_;
```
Comment (slaff) [inline samples/HttpServer_FirmwareUpload/app/application.cpp]: Please don't use underscores as prefix or suffix in a variable, function/method or class names.

File: samples/HttpServer_FirmwareUpload/app/application.cpp [hunk 16]
```diff
@@ -3,10 +3,120 @@
...
+			if(missingHeaderBytes_ > 0) {
+				const size_t chunkSize = std::min(size, missingHeaderBytes_);
+				memcpy(headerPtr_, data, chunkSize);
+				headerPtr_ += chunkSize;
+				missingHeaderBytes_ -= chunkSize;
+				data += chunkSize;
+				size -= chunkSize;
+				consumed += chunkSize;
+				if(missingHeaderBytes_ == 0) {
+					Serial.printf("Receive image for load address 0x%08X, slot starts at 0x%08X\n", header_.loadAddress,
```
Comment (slaff) [inline samples/HttpServer_FirmwareUpload/app/application.cpp]: Please, use `debug_d` for useful debug messages. For example the line above should become:
```
debug_d(("Receive image for load address 0x%08X, slot starts at 0x%08X\n", header_.loadAddress,startAddress_)
```

File: samples/HttpServer_FirmwareUpload/signtool.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python2
```
Comment (slaff) [inline samples/HttpServer_FirmwareUpload/signtool.py]: The README.md must contain brief information how to create signed ROM and how to use it.

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 11 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: Done (Leftover from debugging).

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 12 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: removed (also leftover from debugging)

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 13 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: `uploadStream->close()` triggers the signature validation and therefore must be invoked *before* returning from `onUpdate` to decided if the received image is valid and the slot can be activated.
Did I miss something? According to https://github.com/SmingHub/Sming/blob/develop/docs/source/experimental/signed-ota.rst, the `close` method is the intended place for signature verification.

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 14 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: Separate file it is. 
Moving it to libsodium isn't the right choice, because libsodium offers cryptographic functions, not firmware update functionality.

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 15 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: Fixed.
(FYI: Unlike leading underscores, a trailing underscore does result in a reserved symbol. This  naming convention is quite common to denote private members (e. g. boost))

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 16 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: Fixed.
However, since this is an example, I thought it would be beneficial to show some output during the upgrade, regardless of the debug settings.

File: samples/HttpServer_FirmwareUpload/signtool.py (hunk 17 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/signtool.py]: The README.rst (already present in the folder) now contains the necessary instructions.

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 13 above)
Comment (slaff) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: > uploadStream->close() triggers the signature validation

Not necessarily . It would be better if there is a `validate()` method that triggers the validation and sets the ok flag. Also the ok() method can be renamed to isValid(). The closing of the file is something separate and does not need to be coupled with the validation.  You can move the logic that deletes the first sector if the validation is not correct to the destructor.

> the close method is the intended place for signature verification.

I will change the documentation later in the week.

File: .travis/install.sh [hunk 18]
```diff
@@ -1,6 +1,8 @@
 #!/bin/bash
 set -ex # exit with nonzero exit code if anything fails
 
+pip install --user PyNaCl
```
Comment (slaff) [inline .travis/install.sh]: Concerning the python signing tool: 
- Create a file called requirements.txt that contains all required packages. Similar to the documentation requirements. The file should be in the same directory as the python signing tool.
- Replace the line above with
```
pip install  -r <path-to-signing-tool>/requirements.txt
```

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 14 above)
Comment (slaff) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: > .. Learned a lot today!  ...
> .. but this should be done in a separate PR.

Wonderful! Thanks for the fantastic work so far.

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 13 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: Moved validation code out of the `close` method. Still don't like it, because the `onUpload` callback should inform the user that *all* of the firmware update process has been completed successfully. Even if the signature matches, writing the flash can still fail, and as the last chunk of flash is only written in `RbootOutputStream::close` that must come before preparing the user response. 
I also don't like than an upcast from `ReadWriteStream` to `SignedRbootOutputStream` is necessary in `onUpload`. It may be an acceptable compromise for this example, but for a potential future component, I would prefer a more type-safe solution.

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 14 above)
Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: Thanks for your support as well!

File: .travis/install.sh (hunk 18 above)
Comment (aemseemann) [reply .travis/install.sh]: Used pip3 instead, same as for the doc packages.
However, Travis takes quite some time to install it. Seems like it is building something from source. 
Maybe it has something to do with pip being outdated: "You are using pip version 8.1.1, however version 19.2.1 is available."

**Comment by mikee47** [comment]
May I suggest putting the requirements into the project's component.mk instead?
Plus couple of tags so custom targets show up for `make help`.

```
diff --git a/samples/HttpServer_FirmwareUpload/component.mk b/samples/HttpServer_FirmwareUpload/component.mk
index 74ac73b17..d78808aba 100644
--- a/samples/HttpServer_FirmwareUpload/component.mk
+++ b/samples/HttpServer_FirmwareUpload/component.mk
@@ -6,4 +6,6 @@ COMPONENT_DEPENDS := libsodium
 ENABLE_HTTP_SERVER_MULTIPART = 1
 
+##@Building
+
 web-pack:
 	$(Q) date +'%a, %d %b %Y %H:%M:%S GMT' -u > web/.lastModified
@@ -12,4 +14,7 @@ web-upload: web-pack spiffs-image-update
 	$(call WriteFlash,$(RBOOT_SPIFFS_0)=$(SPIFF_BIN_OUT))
 
+.PHONY: requirements
+requirements:
+	python -m pip install --user -r $(PROJECT_DIR)/requirements.txt
 
 SIGNTOOL := python $(COMPONENT_PATH)/signtool.py
@@ -17,5 +22,5 @@ SIGNING_KEY := $(COMPONENT_PATH)/signing.key
 VERIFICATION_HEADER := $(COMPONENT_PATH)/app/FirmwareVerificationKey.h
 
-$(SIGNING_KEY):	
+$(SIGNING_KEY): requirements
 	@echo "#########################################################"
 	@echo "# Generating new signing key for firmware update images #"
@@ -28,6 +33,5 @@ $(SIGNING_KEY):
 
 .PHONY: generate_signing_key
-generate_signing_key: 
-	@echo Generate a new firmware update signing key on explicit request.
+generate_signing_key: ##Generate a new firmware update signing key on explicit request.
 	$(Q) $(SIGNTOOL) --genkey --keyfile=$(SIGNING_KEY)
 
@@ -44,5 +48,5 @@ $(SIGNED_ROM0): $(RBOOT_ROM_0_BIN) $(SIGNING_KEY)
 
 .PHONY: signedrom
-signedrom: $(SIGNED_ROM0)
+signedrom: $(SIGNED_ROM0) ##Create signed ROM image
 
 
```

File: samples/HttpServer_FirmwareUpload/app/application.cpp (hunk 13 above)
Comment (slaff) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: > I also don't like than an upcast from ReadWriteStream to SignedRbootOutputStream is necessary in onUpload. It may be an acceptable compromise for this example, but for a potential future component, I would prefer a more type-safe solution.

Feel free to suggest better solution as PR. This API is not final.

Comment (aemseemann) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: I'll try to come up with something, but it may take a couple of weeks.

What's your opinion on the current state of the example. Is it acceptable for this PR?

Comment (slaff) [reply samples/HttpServer_FirmwareUpload/app/application.cpp]: > What's your opinion on the current state of the example. Is it acceptable for this PR?

You are doing great and I am happy to welcome you to our team and thank you for your first accepted contribution :)
